File: java/skyline.java
class Solution {
    public int maxIncreaseKeepingSkyline(int[][] grid) {
        int sum = 0;
        for (int i = 0; i < grid[0].length; i++) {
            for (int j = 0; j < grid[0].length; j++) {
                int newHeight = Math.min(maxRow(grid, i), maxCol(grid, j));
                sum += newHeight - grid[i][j];
                grid[i][j] = newHeight;
            }
        }
        return sum;
    }
    
    public int maxRow(int[][] grid, int row) {
        int max = 0;
        for (int i = 0; i < grid.length; i++) {
            if (grid[row][i] > max) max = grid[row][i];
        }
        return max;
    }
    
    public int maxCol(int[][] grid, int col) {
        int max = 0;
        for (int i = 0; i < grid.length; i++) {
            if (grid[i][col] > max) max = grid[i][col];
        }
        return max;
    }
}

File: python/skyline.py
class Solution:
    def maxIncreaseKeepingSkyline(self, grid):
        maxrows, maxcols, result = [], [], 0
        for i in range(len(grid)):
            maxrows.append(self.maxRow(grid, i))
            maxcols.append(self.maxCol(grid, i))
        
        for i in range(len(grid)):
            for j in range(len(grid)):
                result += min(maxrows[i], maxcols[j]) - grid[i][j]
        
        return result
        
    def maxRow(self, grid, rownum):
        return max(grid[rownum])
    def maxCol(self, grid, colnum):
        max = 0
        for i, row in enumerate(grid):
            if row[colnum] > max:
                max = row[colnum]
        return max